core/tests/test-dependencies/route-to-module-mapping-generator.ts

```diff
@@ -0,0 +1,427 @@
+  if (
+    !angularModuleObjectArgument.isKind(ts.SyntaxKind.ObjectLiteralExpression)
+  ) {
+    throw new Error(
+      `The Angular Module decoration in the file: ${routingModuleSourceFile.getFilePath()} does not have an object argument.`
+    );
+  }
+  const importsProperty =
+    angularModuleObjectArgument.getPropertyOrThrow('imports');
+  const importsArray = importsProperty.getFirstChildByKindOrThrow(
+    ts.SyntaxKind.ArrayLiteralExpression
+  );
+  const routerModuleCallExpression = importsArray
+    .getElements()
+    .filter(element => {
+      return (
+        element.getText().includes('RouterModule.forRoot') ||
+        (element.getText().includes('RouterModule.forChild') &&
+          element.getKind() === ts.SyntaxKind.CallExpression)
+      );
+    })[0] as CallExpression;
+  if (routerModuleCallExpression === undefined) {
+    return undefined;
+  }
+  const routesArgument = routerModuleCallExpression.getArguments()[0];
+  if (routesArgument.isKind(ts.SyntaxKind.ArrayLiteralExpression)) {
+    return routesArgument.getElements() as ObjectLiteralExpression[];
+  }
+  if (routesArgument.isKind(ts.SyntaxKind.Identifier)) {
+    const routesVariable =
+      routingModuleSourceFile.getVariableDeclarationOrThrow(
+        routesArgument.getText()
+      );
+    return routesVariable
+      .getInitializerIfKindOrThrow(ts.SyntaxKind.ArrayLiteralExpression)
+      .getElements() as ObjectLiteralExpression[];
+  }
+  return undefined;
+};
+
+/**
+ * Gets the Angular route to module mapping from a routing module file.
+ */
+const getRouteToModuleMappingFromRoutingModule = (
+  routingModuleFilePath: string,
+  parentRoutePath?: string,
+  parentRoutingModuleFilePath?: string
+): Map<Route, string> => {
+  let routeToModuleMapping: Map<Route, string> = new Map();
+  const routingModuleSourceFile = project.addSourceFileAtPath(
+    routingModuleFilePath
+  );
+  const routingModuleRouteNodes = getRouteNodesFromRoutingModuleSourceFile(
+    routingModuleSourceFile
+  );
+  if (routingModuleRouteNodes === undefined) {
+    return routeToModuleMapping;
+  }
+
+  for (const routeNode of routingModuleRouteNodes) {
+    routeToModuleMapping = extendMap(
+      routeToModuleMapping,
+      convertRouteNodeToMap(
+        routeNode,
+        parentRoutePath,
+        parentRoutingModuleFilePath
+      )
+    );
+  }
+
+  return routeToModuleMapping;
+};
+
+/**
+ * Validates the route to module mapping by checking that all
+ * modules defined exist in the codebase.
+ */
+const validateRouteToModuleMapping = (
+  routeToModuleMapping: Map<Route, string>
+): void => {
+  for (const module of routeToModuleMapping.values()) {
+    if (!fs.existsSync(path.resolve(ROOT_DIRECTORY, module))) {
+      throw new Error(
+        `The module: ${module} defined in the manual route to module mapping ` +
+          'does not exist in the codebase. Please remove it from the manual ' +
+          'route to module mapping.'
```

Comment:
Can you pinpoint the exact file or constant?

Reply:
Sorry, don't think that is possible, but if this error shows up, the developer should be able to pinpoint where it is by searching up the module (here I suppose you mean if it is a manual route to module or one defined in the app routing module). Though it will mainly be an outdated manual mapping (which will be resolved soon as webpack is removed) since the app routing should error if a it imports a module that doesn't exist.

Reply:
Why is it not possible?

Reply:
Actually, can you clarify what exactly you are asking for here, sorry. This is called in a separate function which only has knowledge of the modules and if you mean to pinpoint the exact "constant", I don't think its really worth to extend this function. This function is mainly used to validate when a webpack page is removed since it is completely migrated, though I don't know exactly what you mean by pinpoint the exact file or constant; though as noted I feel like this error has enough information for a developer to go remove it.